### ISSUE TITLE
windows pkg not working correctly?

### ISSUE THREAD

**Issue by tganzeboom** [saltstack/salt#24594]
Since I want to update my windows minions from 2014.7.5-2 to 2015.5.0-2 I preferably don't want to do that manually.
However, after generating a new windows repo, doing "salt-run winrepo.genrepo", "salt '*' pkg.refresh_db", it seems that "pkg.install salt" does not seem to work anymore.

salt-master:
           Salt: 2015.5.0
         Python: 2.7.5 (default, Jun 17 2014, 18:11:42)
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.5
           Mako: Not Installed

Client is 2014.7.5-2.

Is there a different way of upgrading windows minions now or am I missing something else?

Also, could you make the saltversion grain more explicit, so that is exactly shows which version I'm dealing with or perhaps another grain showing this?

As an example:

[root@machine ~]# salt machine* grains.item saltversion
machine:
    ----------
    saltversion:
        2014.7.5
[root@machine ~]# salt machine* pkg.version "salt*"
machine:
    ----------
    Salt Minion 2014.7.5-2:
        2014.7.5-2

**Comment by cachedout** [comment]
@twangboy Could you please help out here?

**Comment by twangboy** [comment]
Have you tried using pip to upgrade salt?
http://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.pip.html#upgrading-salt-using-pip

**Comment by twangboy** [comment]
@tganzeboom what version of windows are you using. It looks like your command needs to be something like ```salt '*' pkg.install salt-minion``` for 64bit windows and ```salt '*' pkg.install salt-minion_x86```.

**Comment by tganzeboom** [comment]
@twangboy Windows 2008r2 64 bit. The salt-minion has been put in the winrepo of saltstack as described above. Why should the name matter for pkg.install?

**Comment by twangboy** [comment]
@tganzeboom It wasn't clear from your description that you weren't connected to the internet using the salt-winrepo on github. Could you post the sls file you're using and describe where you're keeping the salt-minion installation executable?

**Comment by tganzeboom** [comment]
@twangboy 
the init.sls file:
```console
[root@saltmaster salt]# cat init.sls
salt:
  2014.1.10:
    installer: 'salt://win/repo/salt/Salt-Minion-2014.1.10-AMD64-Setup.exe'
    full_name: Salt Minion 2014.1.10
    locale: en_US
    reboot: False
    install_flags: ' /S /master=saltmaster'
    uninstaller: 'salt/uninst.exe'
  2014.1.11:
    installer: 'salt://win/repo/salt/Salt-Minion-2014.1.11-AMD64-Setup.exe'
    full_name: Salt Minion 2014.1.11
    locale: en_US
    reboot: False
    install_flags: ' /S /master=saltmaster'
    uninstaller: 'salt/uninst.exe'
  2014.1.13:
    installer: 'salt://win/repo/salt/Salt-Minion-2014.1.13-AMD64-Setup.exe'
    full_name: Salt Minion 2014.1.13
    locale: en_US
    reboot: False
    install_flags: ' /S /master=saltmaster'
    uninstaller: 'salt/uninst.exe'
  2014.7.1:
    installer: 'salt://win/repo/salt/Salt-Minion-2014.7.1-AMD64-Setup.exe'
    full_name: Salt Minion 2014.7.1
    locale: en_US
    reboot: False
    install_flags: ' /S /SD IDOK /master=saltmaster'
    uninstaller: 'salt/uninst.exe'
  2014.7.5:
    installer: 'salt://win/repo/salt/Salt-Minion-2014.7.5-2-AMD64-Setup.exe'
    full_name: Salt Minion 2014.7.5
    locale: en_US
    reboot: False
    install_flags: ' /S /master=saltmaster /minion-name=%COMPUTERNAME%'
    uninstaller: 'C:\salt\uninst.exe'
    uninstall_flags: ' /S'
    refresh: true
  2015.5.0:
    installer: 'salt://win/repo/salt/Salt-Minion-2015.5.0-2-AMD64-Setup.exe'
    full_name: Salt Minion 2015.5.0
    locale: en_US
    reboot: False
    install_flags: ' /S /master=saltmaster /minion-name=%COMPUTERNAME%'
    uninstaller: 'C:\salt\uninst.exe'
    uninstall_flags: ' /S'
    refresh: true
[root@saltmaster salt]#
```
The executable is kept here:
```console
[root@saltmaster salt]# pwd
/srv/salt/base/win/repo/salt
[root@saltmaster salt]# ls
init.sls                               Salt-Minion-2014.1.11-AMD64-Setup.exe  Salt-Minion-2014.7.1-AMD64-Setup.exe    Salt-Minion-2014.7.5-AMD64-Setup.exe
Salt-Minion-2014.1.10-AMD64-Setup.exe  Salt-Minion-2014.1.13-AMD64-Setup.exe  Salt-Minion-2014.7.5-2-AMD64-Setup.exe  Salt-Minion-2015.5.0-2-AMD64-Setup.exe
[root@saltmaster salt]#
```

**Comment by twangboy** [comment]
@tganzeboom I've set up my environment like yours and it is also failing for me. Here's what's happening:

1. The minion downloads the Salt-Minion 2015.5 Setup.exe
2. The minion launches the download
3. The first thing the installer does is stop the Salt-Minion service
4. Stopping the salt-minion service kills all processes started by the service... including the installer

When you said in your initial post that "pkg.install does not seem to work anymore" it would lead me to believe that it has worked in the past. Is that the case?

I also noticed that on the https://github.com/saltstack/salt-winrepo there are .sls files for salt... which would lead me to believe that someone has this working. 

@thebigbear is the one who created those files, perhaps he could shed some light on the subject?

If it does not work, perhaps we need to add an additional option to pkg.install to launch the installer via the task scheduler so that it executes outside of salt. I need to talk to @UtahDave about the best approach here.

**Comment by dariusjs** [comment]
@twangboy when I was evaluating saltstack, I had this functionality working in the past. From memory the 2014.1.x tree was quite ok with this, I was not able to tell by code compare on what has changed. My way around this temporarily was to run this powershell script (it needs powershell 3.x or higher)

...
$saltversion = "Salt-Minion-2015.5.0-AMD64-Setup.exe"

$source = "http://some_internal_host/$saltversion"
$destination = "c:\test\Salt-Minion-2015.5.0-AMD64-Setup.exe"

Set-Location C:\temp
wget $source -OutFile c:\temp\$saltversion

iex 'c:\temp\Salt-Minion-2015.5.0-AMD64-Setup.exe /S /master=saltmaster/minion-name=$env:computername'
...

**Comment by TheBigBear** [comment]
@twangboy and @tganzeboom I can't be 100% sure, but I am _'pretty'_ sure this used to work _'just fine'_. 
But I have been away from my work for the last 3 months and so this has not been watched all that closeley. 
Not quite sure when exactly it 'broke', sorry.

**Comment by tganzeboom** [comment]
@twangboy Thanks for setting up the environment like mine. Care to share your setup of this? I'm interested if you did anything special for the windows part.

**Comment by twangboy** [comment]
@tganzeboom I spun up a Win 2008R2 VM and installed Salt 2014.7.5-2.
Then I set up SLS and config settings to match what you provided. Nothing special.
Like I said, this problem is caused by the fact that the windows installer kills the salt-minion service prior to installation. This kills the salt-minion installation at the same time.

**Comment by tganzeboom** [comment]
@twangboy The Win 2008R2 VM, KVM or VMware or something else?

**Comment by twangboy** [comment]
I'm sorry. It's an instance on Amazon Cloud (AWS). I just refer to it as a VM.

**Comment by twangboy** [comment]
@tganzeboom I'm sorry this has taken so long.
Salt has never officially supported upgrading salt on Windows through the WinRepo. 
I have figured out a work around for you using a combination of @dariusjs 's solution and salt states. Let me know if this works for you or not. We are looking at a few different options for simplifying salt upgrades on Windows. Until then, try this solution.

You'll need 2 things:
- A PowerShell Script (upgrade_salt.ps1)
- A State File (upgrade.sls)

### Powershell Script
On the Salt master, set up a location in the file_roots to store your powershell script file. My file_roots is set up as ```/srv/salt/``` so I created a ```scripts``` folder in that directory.
Create a new file in that directory called ```upgrade_salt.ps1```
The PowerShell script is as follows:
```powershell
# Load parameters
param(
    [string]$salt = "Salt-Minion-2015.5.3-AMD64-Setup.exe",
    [string]$temp = "C:\Temp",
    [string]$master = "",
    [string]$minion = ""
)

New-Item $temp -ItemType Directory -Force

$webclient = New-Object System.Net.WebClient

$source = "http://docs.saltstack.com/downloads/$salt"
$destination = "$temp\$salt"

$webclient.DownloadFile($source, $destination)
If ($master -ne "") {
    $master = "/master=$master"
}
If ($minion -ne "") {
    $minion = "/minion-name=$minion"
}
& $destination /S $master $minion
```
You'll notice from the file above that this script accepts 4 arguments:
- salt : the name of the salt installation file you want to run
- temp : where you want to store the exe and the script
- master : the name or ip address of the master
- minion : the name of the minion
Because this is an upgrade, you don't need to specify master and minion unless you want them to be different than the current master and minion name. If they are not supplied the current settings will be used.

### State File
On the Salt Master create a state file to upgrade salt. I named mine upgrade and put it in a folder called test. It should contain the following content:
```yaml
{% set salt = pillar.get('salt', 'Salt-Minion-2015.5.3-AMD64-Setup.exe') %}
{% set temp = pillar.get('temp', 'C:\Temp') %}
{% set master = pillar.get('master', '') %}
{% set minion = pillar.get('minion', '') %}

{% set args = '-salt ' ~ salt ~ ' -temp ' ~ temp %}

{% if master != '' %}
{%     set args = args ~ ' -master ' ~ master %}
{% endif %}

{% if minion != '' %}
{%     set args = args ~ ' -minion ' ~ minion %}
{% endif %}

copy_script:
  file.managed:
    - name: C:\temp\upgrade_salt.ps1
    - source: salt://scripts/upgrade_salt.ps1

schedule_script:
  cmd.run:
    - name: 'schtasks /RU "System" /Create /TN upgrade-salt /TR "PowerShell -File C:\temp\upgrade_salt.ps1 {{ args }}" /sc ONCE /sd 01/15/1975 /st 01:00 /F /V1 /Z'
    - shell: 'powershell'

run_script:
  cmd.run:
    - name: 'schtasks /Run /TN upgrade-salt'
```
Notice that salt and temp have default values. If you run the state file the minion will be updated to 2015.5.3.

### Upgrade
To upgrade salt-minion run the following command:
```bash
salt 't*64' state.sls test.upgrade
```
This command uses the defaults in the .sls file.

To specify a specific version, send some pillar data like so:
```bash
salt 't*64' state.sls test.upgrade pillar='{"salt": "Salt-Minion-2015.5.3-Test-AMD64-Setup.exe"}'
```

To change the location of the temp directory add some more pillar data:
```bash
salt 't*64' state.sls test.upgrade pillar='{"salt": "Salt-Minion-2015.5.3-Test-AMD64-Setup.exe", "temp": "C:\Windows\Temp"}'
```

To change the master and minion settings add those as well:
```bash
salt 't*64' state.sls test.upgrade pillar='{"salt": "Salt-Minion-2015.5.3-Test-AMD64-Setup.exe", "temp": "C:\Windows\Temp", "master": "192.168.0.12", "minion": "new_minion_name"}'
```
**Note:** You can make all these settings in the .sls file instead of using the command line.

**Note** On older versions of the install, the salt-minion service isn't starting when run as a scheduled task from a state as well as in a few other scenarios. I am in the process of updating the installation files to work correctly. I will post them later this afternoon.
@UtahDave @TheBigBear @dariusjs

**Comment by twangboy** [comment]
This fixes the problem with the service not starting: #25272

Here are the rebuilt packages:
http://docs.saltstack.com/downloads/Salt-Minion-2015.5.3-2-AMD64-Setup.exe
http://docs.saltstack.com/downloads/Salt-Minion-2015.5.3-2-x86-Setup.exe

**Comment by TheBigBear** [comment]
@twangboy this is great, thanks. One comment about the ```'/sd 01/15/1975'``` parameter you use in the sample above. This leads to an 'illegal start date' error message on machines with european date format, for european machines this same date would be written as ```'15/01/1975'```. 
I suggest using somethign like ```'/sd 01/01/1975'``` instead as that is a legal and valid date format anywhere. (I think)

**Comment by twangboy** [comment]
@TheBigBear Brilliant. I'll make the change.

**Comment by twangboy** [comment]
@TheBigBear For other locales the format may be YYYY/MM/DD. Everyone will have to modify the date according to their locale. 01/01/1975 covers most I think.

**Comment by tganzeboom** [comment]
An addition for windows (the remote execute), which I ran into:
```yaml
{% set salt = pillar.get('salt', 'Salt-Minion-2015.5.3-2-AMD64-Setup.exe') %}
{% set temp = pillar.get('temp', 'C:\Temp') %}
{% set master = pillar.get('master', '') %}
{% set minion = pillar.get('minion', '') %}

{% set args = '-salt ' ~ salt ~ ' -temp ' ~ temp %}

{% if master != '' %}
{%     set args = args ~ ' -master ' ~ master %}
{% endif %}

{% if minion != '' %}
{%     set args = args ~ ' -minion ' ~ minion %}
{% endif %}

copy_script:
  file.managed:
    - name: C:\temp\upgrade_salt.ps1
    - source: salt://scripts/upgrade_salt.ps1

remote_execute_on:
  cmd.run:
    - name: 'set-executionpolicy remotesigned'
    - shell: 'powershell'

schedule_script:
  cmd.run:
    - name: 'schtasks /RU "System" /Create /TN upgrade-salt /TR "PowerShell -File C:\temp\upgrade_salt.ps1 {{ args }}" /sc ONCE /sd 01/01/1975 /st 01:00 /F /V1 /Z'
    - shell: 'powershell'

run_script:
  cmd.run:
    - name: 'schtasks /Run /TN upgrade-salt'

#schedule_remote_execute_off:
#  cmd.run:
#    - name: 'schtasks /RU "System" /Create /TN remote-off /TR "PowerShell set-executionpolicy restricted" /sc ONCE /sd 01/01/1975 /st 01:00 /F /V1 /Z'
#    - shell: 'powershell'
```
Just as an afterthought, since version 2015.5.3 is already out, version 2015.5.2 of the minion did not automagically restart after running this script, version 2015.5.3 did, however.

**Comment by TheBigBear** [comment]
@tganzeboom actually, I was really surprised to accidentally find yesterday that the windows 'salt-minion' now updates correctly for 2015.8.rc0 (and even for  2015.5.3-2) wihtout any of the above shenanigans. 

Can it be simply a result of the comment @twangboy made above about " Fixed problem with service not starting #25272 " ? 

Anyway a simple ```salt my-win-minion-1 pkg.install salt-minion``` worked flawless for me yesterday to update from ```'2015.5.3-2'``` to .```'2015.8.rc0'``` and then also the downgrade back to ```'2015.5.3-2'```, worked without a hiccup. 

Do you @tganzeboom ^^^^ get the same ?

**Comment by tganzeboom** [comment]
@TheBigBear Unfortunately not, tried with 2 machines.

It seems to remove the minion, but not reinstall if done by de winrepo way like you descibed above.
